### PR TITLE
[#IR-321]As a user, when I mark a ticket resolved, I should have options to select the reason of resolution

### PR DESCRIPTION
**Pull request by Ravi2714**

### CODE REVIEW

File: apps/retrospective/constants.go [hunk 1]
```diff
@@ -25,3 +25,20 @@ const (
 	GoodRating
 	NotableRating
 )
+
+// Resolution ...
```
Comment (himanshu-garg-165) [inline apps/retrospective/constants.go]: Please move this inside the `task` model file.

File: apps/retrospective/models/task.go [hunk 2]
```diff
@@ -62,13 +65,45 @@ func (task *Task) BeforeUpdate(db *gorm.DB) (err error) {
...
+	task.Meta(&taskResolutionMeta)
 
 	task.IndexAttrs("-SprintMemberTasks")
 	task.NewAttrs("-SprintMemberTasks")
 	task.EditAttrs("-SprintMemberTasks")
 	task.ShowAttrs("-SprintMemberTasks")
 }
+func getTaskResolutionMeta() admin.Meta {
+	return admin.Meta{
+
```
Comment (himanshu-garg-165) [inline apps/retrospective/models/task.go]: Please remove this new line.

File: apps/retrospective/constants.go [hunk 3]
```diff
@@ -25,3 +25,20 @@ const (
...
+type Resolution int8
```
Comment (himanshu-garg-165) [inline apps/retrospective/constants.go]: Make constants for the list of resolution types.

File: apps/retrospective/serializers/base.go [hunk 4]
```diff
@@ -4,3 +4,8 @@ package serializers
 type BaseRating struct {
 	Rating *int8 `json:"Rating" binding:"omitempty,is_valid_rating"`
 }
+
+// BaseResolution is the base serializer for Resolution
+type BaseResolution struct {
```
Comment (himanshu-garg-165) [inline apps/retrospective/serializers/base.go]: Do we really need this?

File: apps/retrospective/serializers/validators/retrospective.go [hunk 5]
```diff
@@ -108,6 +109,24 @@ func IsValidRating(
...
+	v *validator.Validate,
+	topStruct reflect.Value,
+	currentStruct reflect.Value,
+	field reflect.Value,
+	fieldType reflect.Type,
+	fieldKind reflect.Kind,
+	param string,
+) bool {
+	resolution := currentStruct.Interface().(retrospectiveSerializers.BaseResolution).Resolution
+	if resolution != nil && *resolution >= 1 && int(*resolution) < len(retrospective.ResolutionValues) {
```
Comment (himanshu-garg-165) [inline apps/retrospective/serializers/validators/retrospective.go]: Use a constant for `1`.

File: apps/retrospective/services/sprint_task.go [hunk 6]
```diff
@@ -257,6 +258,7 @@ func (service SprintTaskService) MarkUndone(
 		Where("tasks.deleted_at IS NULL").
 		Where("id = (?)", query).
 		Update("done_at", nil).
+		Updates(map[string]interface{}{"done_at": nil, "resolution": nil}).
```
Comment (himanshu-garg-165) [inline apps/retrospective/services/sprint_task.go:261]: It should be either this,
```suggestion
		Updates("resolution": nil).
```
...or remove the line above.

File: apps/retrospective/services/sprint_task.go [hunk 7]
```diff
@@ -370,7 +372,8 @@ func (service SprintTaskService) tasksWithTimeDetailsForCurrentAndPrevSprint(ret
 						sprint_current_members.sprint_current_member_time as sprint_current_member_time,
             tasks.estimate,
             tasks.rating,
-            tasks.done_at,
+						tasks.done_at,
```
Comment (himanshu-garg-165) [inline apps/retrospective/services/sprint_task.go:375]: Same for below:
```suggestion
            tasks.done_at,
```

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/retrospective/constants.go (hunk 1 above)
Comment (Ravi2714) [reply apps/retrospective/constants.go]: done

File: apps/retrospective/models/task.go (hunk 2 above)
Comment (Ravi2714) [reply apps/retrospective/models/task.go]: done

File: apps/retrospective/constants.go (hunk 3 above)
Comment (Ravi2714) [reply apps/retrospective/constants.go]: done

File: apps/retrospective/serializers/base.go (hunk 4 above)
Comment (Ravi2714) [reply apps/retrospective/serializers/base.go]: removed

File: apps/retrospective/serializers/validators/retrospective.go (hunk 5 above)
Comment (Ravi2714) [reply apps/retrospective/serializers/validators/retrospective.go]: done

File: apps/retrospective/services/sprint_task.go (hunk 6 above)
Comment (Ravi2714) [reply apps/retrospective/services/sprint_task.go:261]: done

File: apps/retrospective/services/sprint_task.go (hunk 7 above)
Comment (Ravi2714) [reply apps/retrospective/services/sprint_task.go:375]: changed

**Review by himanshu-garg-165 — APPROVED** [review]

**Pull request merged**